apollo-router-core/src/traits.rs

```diff
@@ -4,6 +4,15 @@ use futures::prelude::*;
 use std::sync::Arc;
 use std::{fmt::Debug, pin::Pin};
 
+/// A cache resolution trait.
+///
+/// Clients of CachingMap are required to provider a resolver during Map creation. The resolver
+/// will be used to find values for cache misses. A Result is expected, because retrieval may fail.
+#[async_trait]
+pub trait CacheResolver<K, V> {
+    async fn retrieve(&self, key: K) -> Result<V, CacheResolverError>;
+}
```

Comment:
I think we don't need this trait. You can just use `Fn(K) -> Result<V, E>`

Reply:
I tried a few different approaches around using a function callback and all of them looked more complicated than using the resolver trait. I can try poking it around again a bit more.

Reply:
I think this conversation has moved into the other comment thread, so I'll make this as "resolved" to keep the conversation focussed.